### DISCUSSION TITLE
Getting "Attempted to spawn an Actor outside of a service" error

### DISCUSSION THREAD

**Discussion by pipex** [statelyai/xstate#1295] [General] [answered]
Hi, I posted something related to this in spectrum but also adding it here. 

In my company we are evaluating the use of xstate and I'm working on a proof of concept for controlling an embedded device state using xstates (with actors). 

I'm getting the `Attempted to spawn an Actor outside of a service` error when spawning a child but not sure what I'm doing wrong.

I replicated the problem with two actors in a sandbox here. 

https://codesandbox.io/s/admiring-paper-blktf?file=/src/index.ts

If anyone has any suggestions I'd love to hear them

**Comment by davidkpiano** [comment, marked as answer]
Working on a fix for this! https://github.com/davidkpiano/xstate/pull/1290

It should be an innocuous warning, but it will be solved when that PR goes in.

**Reply by pipex** [reply]
Thanks @davidkpiano, I also just realised there is a bug in my example

For spawning a child I was doing
```
actions: assign({ ponger: spawn(ponger) })
```

But it should be
```
actions: assign({ ponger: () => spawn(ponger) })
```

Awesome work with the framework btw, I think is very powerful and an awesome way to bring model-based design into modern development.

**Reply by davidkpiano** [reply]
The (tentative) plan for V5 will be to have spawned entities be lazy by default, so that this would be possible:

```js
actions: assign({ ponger: spawn(() => ponger) });

// This is also okay
actions: assign({ ponger: () => spawn(() => ponger) });
```

And this should never occur (will throw an error):

```js
// none of these are lazy spawns
actions: assign({ ponger: spawn(ponger) }); // ❌
actions: assign({ ponger: () => spawn(ponger) }); // ❌
```

This is not implemented yet in V5, but will be.

**Reply by pipex** [reply]
Another question, somewhat related. I see that the [spawn function](https://xstate.js.org/api/globals.html#spawn) is overloaded to return an Interpreter or an actor. 

But when I do 
```javascript
actions: assign({ ponger: () => spawn(ponger) })
```

The typescript compiler usually complains with the warning `Type 'Actor<PongContext, PongEvent>' is missing the following properties from type 'Interpreter<PongContext, any, PongEvent, any>'`.

Can you think of any way to coerce the assignment to prevent this?

**Reply by davidkpiano** [reply]
In V5 these will always be `ActorRef` types (same as an actor, don't worry about the difference for now), but have you tried `as Actor`?

**Reply by pipex** [reply]
🤦🏻‍♂️ was trying something a lot more complicated, thanks again
